src/skagent/rule_module.py

```diff
@@ -0,0 +1,278 @@
+    elif isinstance(rule, str):
+        # String expression
+        deps = _TOKEN_RE.findall(rule)
+    elif callable(rule):
+        # Callable function
+        try:
+            deps = list(inspect.signature(rule).parameters.keys())
+        except Exception:
+            # Fallback: parse source code
+            try:
+                src = inspect.getsource(rule)
+                deps = _TOKEN_RE.findall(src)
+            except Exception:
+                pass
+
+    return deps
+
+
+def format_callable_rule(var, rule):
+    """
+    Format a callable rule (function/lambda) into a string.
+
+    Parameters
+    ----------
+    var : str
+        The variable name
+    rule : callable
+        The callable rule
+
+    Returns
+    -------
+    str
+        Formatted rule string
+    """
+    try:
+        src = inspect.getsource(rule).strip()
+        if "lambda" in src:
+            # Extract lambda body
+            body = src.split(":", 1)[1].strip()
+            # Remove trailing comma or parenthesis
+            body = body.rstrip(",)")
+            return f"{var} = {body}"
+        else:
+            return f"{var} = [Function]"
+    except Exception:
+        return f"{var} = [Unknown]"
+
+
+def format_control_rule(var, control):
+    """
+    Format a Control rule into a string.
+
+    Parameters
+    ----------
+    var : str
+        The variable name
+    control : Control
+        The Control object
+
+    Returns
+    -------
+    str
+        Formatted Control rule string
+    """
+    deps = sorted(control.iset)
+    bounds_info = []
+    if control.lower_bound:
+        bounds_info.append("lower_bound")
+    if control.upper_bound:
+        bounds_info.append("upper_bound")
+    bounds_str = f", {', '.join(bounds_info)}" if bounds_info else ""
+    return f"{var} = Control({', '.join(deps)}{bounds_str})"
+
+
+def format_string_rule(var, expression):
+    """
+    Format a string expression rule.
+
+    Parameters
+    ----------
+    var : str
+        The variable name
+    expression : str
+        The mathematical expression string
+
+    Returns
+    -------
+    str
+        Formatted rule string
+    """
+    return f"{var} = {expression}"
+
+
+def format_constant_rule(var, value):
+    """
+    Format a constant value rule.
+
+    Parameters
+    ----------
+    var : str
+        The variable name
+    value : any
+        The constant value
+
+    Returns
+    -------
+    str
+        Formatted rule string
+    """
+    return f"{var} = {value}"
+
+
+def validate_rule(rule):
+    """
+    Validate that a rule is well-formed.
+
+    Parameters
+    ----------
+    rule : any
+        The rule to validate
+
+    Returns
+    -------
+    bool
+        True if rule is valid
+
+    Raises
+    ------
+    ValueError
+        If rule is invalid
+    """
+    if isinstance(rule, Control):
+        if not rule.iset:
+            raise ValueError("Control rule must have non-empty information set")
```

Comment:
That is not well worded. It is possible for a Control to have an empty information set. That is actually a feature we want to build in #67. But it must specific an information set.